client/src/telethon/_impl/mtsender/sender.py

```diff
@@ -234,50 +242,55 @@ async def _step_until_receive(self, rx: Future[bytes]) -> bytes:
             if rx.done():
                 return rx.result()
 
-    async def step(self) -> list[Updates]:
-        async with self.lock:
-            return await self._step()
+    async def get_updates(self) -> list[Updates]:
+        await self.step()
+        result = self._updates.copy()
+        self._updates.clear()
+        return result
 
-    async def _step(self) -> list[Updates]:
-        self._try_fill_write()
+    async def step(self) -> None:
+        self._step_event.clear()
 
-        recv_req = asyncio.create_task(self._request_event.wait())
-        recv_data = asyncio.create_task(self._reader.read(MAXIMUM_DATA))
-        send_data = asyncio.create_task(self._do_send())
-        done, pending = await asyncio.wait(
-            (recv_req, recv_data, send_data),
-            timeout=self._next_ping - asyncio.get_running_loop().time(),
-            return_when=FIRST_COMPLETED,
-        )
+        if not self._recv_lock.locked():
+            asyncio.create_task(self.step_recv())
+        if not self._send_lock.locked():
+            asyncio.create_task(self.step_send())
 
-        if pending:
-            for task in pending:
-                task.cancel()
-            await asyncio.wait(pending)
+        await self._step_event.wait()
 
-        result = []
-        if recv_req in done:
-            self._request_event.clear()
-        if recv_data in done:
-            result = self._on_net_read(recv_data.result())
-        if send_data in done:
-            self._on_net_write()
-        if not done:
+    async def step_recv(self) -> None:
+        async with self._recv_lock:
+            try:
+                await self._step_recv()
+            finally:
+                self._step_event.set()
```

Comment:
step_recv and step_send can fire once when step is called, otherwise self._step_event.wait() will wait. If the step is completed in any way, the other tasks will also be completed